src/t_set.c

```diff
@@ -63,81 +64,96 @@ static void maybeConvertIntset(robj *subject) {
 /* Add the specified sds value into a set.
  *
  * If the value was already member of the set, nothing is done and 0 is
- * returned, otherwise the new element is added and 1 is returned.
- *
- * If you don't have an sds value, use setTypeAddBuf() instead. */
+ * returned, otherwise the new element is added and 1 is returned. */
 int setTypeAdd(robj *subject, sds value) {
-    if (subject->encoding == OBJ_ENCODING_HT) {
-        /* Specialized code to avoid creating a new sds in some case. */
-        dict *ht = subject->ptr;
-        dictEntry *de = dictAddRaw(ht,value,NULL);
-        if (de) {
-            dictSetKey(ht,de,sdsdup(value));
-            dictSetVal(ht,de,NULL);
-            return 1;
-        } else {
-            return 0;
+    return setTypeAddAux(subject, value, sdslen(value), 0, 1);
+}
+
+/* Add member. This function is optimized for the different encodings. The
+ * value can be provided as an sds string (indicated by passing str_is_sds =
+ * 1), as string and length (str_is_sds = 0) or as an integer in which case str
+ * is set to NULL and llval is provided instead. */
+int setTypeAddAux(robj *set, char *str, size_t len, int64_t llval, int str_is_sds) {
+    char tmpbuf[LONG_STR_SIZE];
+    if (!str) {
+        if (set->encoding == OBJ_ENCODING_INTSET) {
+            uint8_t success = 0;
+            set->ptr = intsetAdd(set->ptr, llval, &success);
+            if (success) maybeConvertIntset(set);
+            return success;
         }
+        /* Convert int to string. */
+        len = ll2string(tmpbuf, sizeof tmpbuf, llval);
+        str = tmpbuf;
+        str_is_sds = 0;
```

Comment:
Right, there is already `lpAppendInteger`. I'll use it.

Reply:
We need lpFindInteger before using lpAppendInteger. I'll leave this out for now.

Reply:
I did a "miniptimization", using lpAppendInteger instead of lpAppend if str == tmpbuf (which means we had an integer originally but converted it to string for lpFind).